engine/src/main/java/org/pentaho/di/repository/kdr/delegates/metastore/KettleDatabaseRepositoryMetaStore.java

```diff
@@ -0,0 +1,460 @@
+import org.pentaho.metastore.api.exceptions.MetaStoreException;
+import org.pentaho.metastore.api.exceptions.MetaStoreNamespaceExistsException;
+import org.pentaho.metastore.api.security.IMetaStoreElementOwner;
+import org.pentaho.metastore.api.security.MetaStoreElementOwnerType;
+import org.pentaho.metastore.stores.memory.MemoryMetaStore;
+import org.pentaho.metastore.stores.memory.MemoryMetaStoreElementOwner;
+
+public class KettleDatabaseRepositoryMetaStore extends MemoryMetaStore implements IMetaStore {
+
+  protected KettleDatabaseRepository repository;
+  private KettleDatabaseRepositoryMetaStoreDelegate delegate;
+
+  public KettleDatabaseRepositoryMetaStore( KettleDatabaseRepository repository ) {
+    this.repository = repository;
+    delegate = repository.metaStoreDelegate;
+  }
+
+  // Handle namespaces...
+
+  @Override
+  public List<String> getNamespaces() throws MetaStoreException {
+
+    try {
+      List<String> namespaces = new ArrayList<String>();
+      Collection<RowMetaAndData> namespaceRows = delegate.getNamespaces();
+      for ( RowMetaAndData namespaceRow : namespaceRows ) {
+        String namespace = namespaceRow.getString( KettleDatabaseRepository.FIELD_NAMESPACE_NAME, null );
+        if ( !Utils.isEmpty( namespace ) ) {
+          namespaces.add( namespace );
+        }
+      }
+      return namespaces;
+
+    } catch ( Exception e ) {
+      throw new MetaStoreException( e );
+    }
+  }
+
+  @Override
+  public void createNamespace( String namespace ) throws MetaStoreException, MetaStoreNamespaceExistsException {
+    try {
+      ObjectId namespaceId = delegate.getNamespaceId( namespace );
+      if ( namespaceId != null ) {
+        throw new MetaStoreNamespaceExistsException( "Namespace with name '" + namespace + "' already exists" );
+      }
+
+      // insert namespace into R_NAMESPACE
+      //
+      delegate.insertNamespace( namespace );
+      repository.commit();
+    } catch ( Exception e ) {
+      repository.rollback();
+      throw new MetaStoreException( e );
+    }
+  }
+
+  @Override
+  public void deleteNamespace( String namespace ) throws MetaStoreException, MetaStoreDependenciesExistsException {
+    try {
+      ObjectId namespaceId = delegate.verifyNamespace( namespace );
+
+      List<IMetaStoreElementType> elementTypes = getElementTypes( namespace );
+      if ( !elementTypes.isEmpty() ) {
+
+        List<String> dependencies = new ArrayList<String>();
+        for ( IMetaStoreElementType elementType : elementTypes ) {
+          dependencies.add( elementType.getId() );
+        }
+        throw new MetaStoreDependenciesExistsException( dependencies, "The namespace to delete, '"
+          + namespace + "' is not empty" );
+      }
+
+      // Now delete the namespace
+      //
+      delegate.deleteNamespace( namespaceId );
+      repository.commit();
+    } catch ( MetaStoreDependenciesExistsException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)